### PR TITLE
More STD variants

### PR DESCRIPTION
**Pull request by tom91136**
This PR adds the index-oriented STD implementation (`std-indices`), it's been verified to work with NVHPC 21.9 and the usual C++ compilers with TBB.

This PR also refactors the STD and STD20 implementation to `std-data` and `std-ranges` respectively. 

Currently this is still pending validation with oneDPL but it's ready for high-level reviews; don't merge yet.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -34,9 +34,9 @@ endmacro()
 #set(CUDA_ARCH sm_70)
 #set(BLT_DIR /home/tom/Downloads/blt-0.3.6/)
 
-#set(MODEL STD)
-#set(ARCH  cc70)
-#set(CXX_EXTRA_FLAGS -v)
+#set(MODEL std-data)
+#set(CMAKE_CXX_COMPILER /home/tom/Downloads/nvhpc_2021_219_Linux_x86_64_cuda_multi/install_components/Linux_x86_64/21.9/compilers/bin/nvc++)
```
Comment (tomdeakin) [inline CMakeLists.txt:38]: Can you just delete these types of comments for the final merge please?

File: src/std-indices/STDIndicesStream.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
```
Comment (tomdeakin) [inline src/std-indices/STDIndicesStream.cpp:1]: Do you need to change this to UoB?

File: src/std-indices/STDIndicesStream.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#pragma once
+
+#include <iostream>
+#include <stdexcept>
+#include "Stream.h"
+
+#define IMPLEMENTATION_STRING "STD (index-oriented)"
+
+template <typename N>
+class ranged {
```
Comment (tomdeakin) [inline src/std-indices/STDIndicesStream.h:15]: Can you add some high level comments explaining what this is? I.e it’s a lightweight counting iterator which will be used by the STL algorithms …..

**Review by tomdeakin — CHANGES_REQUESTED** [review]
Looks great on the whole @tom91136 ! Just a few minor things to add along with the oneDPL.

I wonder if we also need to add comments for each version explaining data-parallel vs index-parallel? Maybe put this in a comment at the top of the `.cpp` files.

**Comment by tom91136** [comment]
I've reverted the oneDPL commit for now (there's already a branch that uses pointers based on this branch so force pushing is probably gonna break stuff)

File: src/std-data/STDDataStream.h [hunk 4]
```diff
@@ -1,4 +1,5 @@
-// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+// Copyright (c) 2020 Tom Deakin
```
Comment (tomdeakin) [inline src/std-data/STDDataStream.h:1]:
```suggestion
// Copyright (c) 2020 NVIDIA CORPORATION. All rights reserved.
```

File: src/std-data/STDDataStream.h [hunk 5]
```diff
@@ -1,4 +1,5 @@
...
+// University of Bristol HPC
```
Comment (tomdeakin) [inline src/std-data/STDDataStream.h:2]:
```suggestion
```

File: src/std-data/STDDataStream.h [hunk 6]
```diff
@@ -1,4 +1,5 @@
-// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+// Copyright (c) 2020 NVIDIA CORPORATION. All rights reserved.
+// University of Bristol HPC
```
Comment (tomdeakin) [inline src/std-data/STDDataStream.h]:
```suggestion
// Copyright (c) 2021 University of Bristol HPC
```

File: src/std-data/STDDataStream.cpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2020 Tom Deakin
```
Comment (tomdeakin) [inline src/std-data/STDDataStream.cpp:1]:
```suggestion
// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
// Updated in 2021 by University of Bristol
```

**Pull request merged**